### PR TITLE
[VER-1123] Expose Wasm custom sections in the state tree

### PR DESCRIPTION
**Pull request by jwiegley**

### CODE REVIEW

File: src/IC/Ref.hs [hunk 1]
```diff
@@ -376,6 +377,19 @@ authReadStateRequest t ecid ev (ReadStateRequest user_id paths) = do
         assertEffectiveCanisterId ecid (EntityId cid)
       ("canister":cid:"controllers":_) ->
         assertEffectiveCanisterId ecid (EntityId cid)
+      ("canister":cid:"metadata":name:_) -> do
+        assertEffectiveCanisterId ecid (EntityId cid)
+        ic <- get
```
[the remark this answers is not in the document]
Comment (nomeata) [reply src/IC/Ref.hs:382]: But, more prominently, anyone is allowed to query these non-existent paths, right? Like this part of https://sdk.dfinity.org/docs/interface-spec/index.html#http-read-state where it is ok for anyone to access a non-existing `request_status`, and only when it does exist some additional access control kicks in.

The alternative would be to make it an authorization error if the canister does not exist (as if all metadata is private)…

[the remark this answers is not in the document]
Comment (chenyan-dfinity) [reply src/IC/Ref.hs:382]: > it is ok for anyone to access a non-existing request_status, and only when it does exist some additional access control kicks in.

Yes

> The alternative would be to make it an authorization error if the canister does not exist

What's the benefit? If the canister is empty, it's equivalent to a non-existed path, right?

[the remark this answers is not in the document]
Comment (nomeata) [reply src/IC/Ref.hs:382]: It always kinda felt wrong that “non-existing” means “access allowed”, but nevermind, I can’t quite put my finger on it right now.

File: src/IC/Canister.hs [hunk 2]
```diff
@@ -1,99 +1,133 @@
...
+    update_methods :: MethodName ↦ (EntityId -> Env -> NeedsToRespond -> Cycles -> Blob -> UpdateFunc),
+    query_methods :: MethodName ↦ (EntityId -> Env -> Blob -> QueryFunc),
+    callbacks :: Callback -> Env -> NeedsToRespond -> Cycles -> Response -> Cycles -> UpdateFunc,
+    cleanup :: WasmClosure -> Env -> WasmState -> TrapOr (WasmState, ()),
+    pre_upgrade_method :: WasmState -> EntityId -> Env -> TrapOr (CanisterActions, Blob),
+    post_upgrade_method :: EntityId -> Env -> Blob -> Blob -> TrapOr (WasmState, CanisterActions),
+    inspect_message :: MethodName -> EntityId -> Env -> Blob -> WasmState -> TrapOr (),
+    heartbeat :: Env -> WasmState -> TrapOr (WasmState, ([MethodCall], CanisterActions)),
+    public_custom_sections :: Text ↦ Blob,
+    private_custom_sections :: Text ↦ Blob
```
[the remark this answers is not in the document]
Comment (jwiegley) [reply src/IC/Canister.hs:54]: I like it, will change it soon.

**Comment by nomeata** [comment]
I  couldn’t help it and thought some more about how to elegantly implement and test tonight, and I’ll propose an alternative implementation in a separate PR – that’s easier than code review ping pong.

**Comment by nomeata** [comment]
How about this? https://github.com/dfinity/ic-hs/pull/55

**Comment by marcin-dziadus** [comment]
@nomeata I guess we can close this PR - the feature is already implemented elsewhere.

**Pull request closed without merging**